### PR TITLE
bridge: add vlan trunk support

### PR DESCRIPTION
**Pull request by tjjh89017**
add vlan trunk support for veth
vlan trunk only support L2 only mode without any IPAM
refer ovs-cni design
https://github.com/k8snetworkplumbingwg/ovs-cni/blob/main/pkg/plugin/plugin.go

design:
origin "vlan" option will be PVID or untagged vlan for the network.
"vlanTrunk" will setup tagged vlan for veth.

entry type:
`{ "id": 100 }` will specify only tagged vlan 100
`{ "minID": 100, "maxID": 120 }` will specify tagged vlan from 100 to
120 (include 100 and 120)
vlanTrunk is a list of above entry type, so you can use this to add
tagged vlan
`[
  { "id": 100 },
  {
    "minID": 1000,
    "maxID": 2000
  }
]`

complete config will be like this
```json
{
  "cniVersion": "0.3.1",
  "name": "mynet",
  "type": "bridge",
  "bridge": "mynet0",
  "vlan": 100,
  "vlanTrunk": [
    { "id": 101 },
    { "minID": 1000, "maxID": 2000 },
    { "minID": 3000, "maxID": 4000 }
  ],
  "ipam": {}
}
```

Original PR: https://github.com/containernetworking/plugins/pull/689

Signed-off-by: Date Huang <date.huang@suse.com>

### CODE REVIEW

**Comment by tjjh89017** [comment]
@squeed Please use this PR for https://github.com/containernetworking/plugins/pull/689
Thanks a lot

**Comment by squeed** [comment]
hey @SchSeba would you mind taking a quick look?

**Comment by tjjh89017** [comment]
I found the test failed in "portmap" plugin.

**Comment by SchSeba** [comment]
I will take a look on this one.
also @mlguerrero12 if you have time please take a look :)

**Comment by tjjh89017** [comment]
Hi @squeed and @SchSeba 

I tried to fix up all lint issue and rebase to current main branche.
Could you help us to look on this PR?
Thanks a lot

File: plugins/main/bridge/bridge.go [hunk 1]
```diff
@@ -127,6 +142,59 @@ func loadNetConf(bytes []byte, envArgs string) (*NetConf, string, error) {
 	return n, n.CNIVersion, nil
 }
 
+// This method is copied from https://github.com/k8snetworkplumbingwg/ovs-cni/blob/v0.27.2/pkg/plugin/plugin.go
+func collectVlanTrunk(vlanTrunk []*VlanTrunk) ([]int, error) {
+	if vlanTrunk == nil {
+		return nil, nil
+	}
+
+	vlanMap := make(map[int]bool)
```
Comment (maiqueb) [inline plugins/main/bridge/bridge.go:151]: might as well use a `map[int]struct{}` which I think is more memory efficient. IIUC, you're using a set (you only care about the keys, and never put the value `false`  in the map).

I also think it is more idiomatic, since the type clearly shows there's nothing in the value to be looked at.

File: plugins/main/bridge/bridge.go [hunk 2]
```diff
@@ -127,6 +142,59 @@ func loadNetConf(bytes []byte, envArgs string) (*NetConf, string, error) {
...
+		}
+		if minID > 0 && maxID > 0 {
+			for v := minID; v <= maxID; v++ {
+				vlanMap[v] = true
+			}
+		}
+
+		// single vid
+		if item.ID != nil {
+			ID = *item.ID
```
Comment (maiqueb) [inline plugins/main/bridge/bridge.go:179]: nit: IMO, there's no need to have the `ID` variable.

File: plugins/main/bridge/bridge.go [hunk 3]
```diff
@@ -127,6 +142,59 @@ func loadNetConf(bytes []byte, envArgs string) (*NetConf, string, error) {
...
+		}
+	}
+
+	if len(vlanMap) == 0 {
+		return nil, nil
+	}
+	vlans := make([]int, 0, len(vlanMap))
+	for k := range vlanMap {
+		vlans = append(vlans, k)
+	}
```
Comment (maiqueb) [inline plugins/main/bridge/bridge.go:193]: I may be missing something, but I don't understand why you first compute a set of the VLAN IDs, and then compute a slice from it.

Couldn't you have just created the slice directly ? And append the trunk IDs to it?

File: plugins/main/bridge/bridge.go [hunk 4]
```diff
@@ -452,6 +531,10 @@ func cmdAdd(args *skel.CmdArgs) error {
 		return fmt.Errorf("cannot set hairpin mode and promiscuous mode at the same time")
 	}
 
+	if n.vlans != nil && len(n.vlans) > 0 && isLayer3 {
```
Comment (maiqueb) [inline plugins/main/bridge/bridge.go:534]: nit: I think `if n.vlans != nil && len(n.vlans) > 0` is redundant; len can handle a nil slice.

**Review by maiqueb — COMMENTED** [review]
some nits.

File: plugins/main/bridge/bridge.go (hunk 2 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:179]: Just copy out and avoid to dereference `item.ID` every time.
If needed, I could remove ID here.

File: plugins/main/bridge/bridge.go (hunk 3 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:193]: I copy those from OVS cni.
IIRC, we want to deduplicate vlan ids and sort them.
If go netlink supports this [PR](https://github.com/vishvananda/netlink/pull/863) and we could call netlink once for contiguous vlan id.

File: plugins/main/bridge/bridge.go (hunk 4 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go]: I will fix this part.

File: plugins/main/bridge/bridge.go (hunk 1 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go]: ok got it.

File: plugins/main/bridge/bridge.go (hunk 3 above)
Comment (mlguerrero12) [reply plugins/main/bridge/bridge.go:193]: I think it is with the aim to support config like:
{ "minID": 100, "maxID": 200 }
{ "minID": 150, "maxID": 200 }

if you use a slice directly, it will append ids 150 to 200 again.

Another thing, the config { "maxID": 120 } does not affect the code (if in 171 addresses this), but shouldn't we return an error saying that maxID was set but not minID?

File: plugins/main/bridge/bridge.go [hunk 5]
```diff
@@ -386,6 +454,14 @@ func setupVeth(netns ns.NetNS, br *netlink.Bridge, ifName string, mtu int, hairp
 		}
 	}
 
+	// S1031: unnecessary nil check around range (gosimple)
```
Comment (mlguerrero12) [inline plugins/main/bridge/bridge.go:457]: no need to add this comment I'd say

File: plugins/main/bridge/bridge.go [hunk 6]
```diff
@@ -386,6 +454,14 @@ func setupVeth(netns ns.NetNS, br *netlink.Bridge, ifName string, mtu int, hairp
...
+	for _, v := range vlans {
```
Comment (mlguerrero12) [inline plugins/main/bridge/bridge.go:458]: the default vlan will be present if the regular vlan is not set and preserveDefaultVlan is true. I would add this inside the if above (adding || len(vlans) != 0 and a if for the BridgeVlanAdd) to avoid this.

File: plugins/main/bridge/bridge.go (hunk 3 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:193]: > Another thing, the config { "maxID": 120 } does not affect the code (if in 171 addresses this), but shouldn't we return an error saying that maxID was set but not minID?

That is a problem, let me think how to deal with it.
Thanks for review!

File: plugins/main/bridge/bridge.go (hunk 5 above)
Comment (maiqueb) [reply plugins/main/bridge/bridge.go:457]: I think this is to appease the linter.

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:457]: I just do a note to make sure no one will add a nil check there again.
I could remove it later

File: plugins/main/bridge/bridge.go (hunk 6 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:458]: I will suggest to move the preserveDefaultVlan condition to upper layer (not in if vlanID != 0 condition).


FROM: 
```
    if vlanID != 0 {
        if !preserveDefaultVlan {
            err = removeDefaultVlan(hostVeth)
            if err != nil {
                return nil, nil, fmt.Errorf("failed to remove default vlan on interface %q: %v", hostIface.Name, err)
            }
        }

        err = netlink.BridgeVlanAdd(hostVeth, uint16(vlanID), true, true, false, true)
        if err != nil {
            return nil, nil, fmt.Errorf("failed to setup vlan tag on interface %q: %v", hostIface.Name, err)
        }
    }
```

TO:
```
    if !preserveDefaultVlan {
        err = removeDefaultVlan(hostVeth)
        if err != nil {
            return nil, nil, fmt.Errorf("failed to remove default vlan on interface %q: %v", hostIface.Name, err)
        }
    }

    if vlanID != 0 {
        err = netlink.BridgeVlanAdd(hostVeth, uint16(vlanID), true, true, false, true)
        if err != nil {
            return nil, nil, fmt.Errorf("failed to setup vlan tag on interface %q: %v", hostIface.Name, err)
        }
    }
```

File: plugins/main/bridge/bridge.go (hunk 5 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:457]: ok, i removed it.

File: plugins/main/bridge/bridge.go (hunk 3 above)
Comment (mlguerrero12) [reply plugins/main/bridge/bridge.go:194]: You can add the contents of the first if as an else of the second one. You don't need to check again if item.MinID or item.MaxID are != nil again.

File: plugins/main/bridge/bridge.go (hunk 6 above)
Comment (mlguerrero12) [reply plugins/main/bridge/bridge.go:459]: the reasoning is that we only modify the default vlan when the vlan filtering is enabled (line 494).

File: plugins/main/bridge/bridge.go (hunk 3 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:194]: The problem is we have 3 case here
1. minID == nil && maxID == nil; need to check ID
2. (minID != nil && maxID == nil) || (minID == nil && maxID != nil); fail case; L174 try to capture.
3. minID != nil && maxID != nul; success case

L157 only capture case 3, so I need to use L174 to check case 1 or 2.
If case 2, and we return error.
otherwise, we should use the condition statement in case 2 in the code

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:192]: I fix this.
I had wrong statement.
current `else if` should be correct.

**Comment by mlguerrero12** [comment]
I see a conflict with PVID here that has not been addressed.

When a port has PVID set (regular vlan id here), this becomes an access port. On the other hand, a trunk port has no PVID. Only tagged VIDs and no VID (default vlan). That's a check we need to add on this pr.

So, a use case for vlan trunk is a container with a trunk port on the bridge. This container can use multus for example to create additional vlan subinterfaces. Let's say two for vlan 100 and 200. Then, 3 containers can be created, one with vlan 100, the other one with 200 and the last one with no vlan. The subinterfaces on the first container should be able to talk to the containers with the specific vlan and untagged traffic should be forwarded thanks to the default vlan.

Now, regarding the default vlan, with the new parameter preserveDefaultVlan, it is possible to remove it. For instance, if a user only wants tagged traffic to go through the trunk port.

Hope this makes sense.

File: plugins/main/bridge/bridge.go (hunk 3 above)
Comment (mlguerrero12) [reply plugins/main/bridge/bridge.go:192]: if item.MinID != nil || item.MaxID != nil {
   print error
} else {
  do work
}

this is what I meant

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:192]: So put `single vid work` in the else statement?

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:192]:
```
if item.MinID != nil || item.MaxID != nil {
print error
} else {
//do work
// add single vid 
}
```

or 

```
if item.MinID != nil || item.MaxID != nil {
print error
} else {
//do work
// add vlan range from minID to maxID
}
```

**Comment by tjjh89017** [comment]
> Exactly, the pvid here is the native vlan, not the one coming from the parameter vlan. From you message, I think you believe the native vlan is configured with this parameter and it is not. We currently don't have a param to modify the native vlan of the bridge.
> 
> When we add the pvid (vlan param equal 100), the port has: 1 Egress untagged 100 PVID Egress untagged
> 
> Now, if you add trunk port. This will become.
> 
> 1 Egress untagged 100 PVID Egress untagged 200 300
> 
> Native vlan is still 1 (this a configuration of the bridge itself, not the ports of the bridge).
> 
> With the above config, untagged ingress traffic will have tag 100 (clearly not the native vlan) and on egress, vid 100 and untagged traffic will output untagged.
> 
> I'm familiar with cisco switches as well and I don't remember a config to do this.

So did it mean that we need to always remove the vid 1 when vlan is presented?

```
When we add the pvid (vlan param equal 100), the port has:
1 Egress untagged
100 PVID Egress untagged
```
In this config, vid 1 didn't take any effect and it might have some issue on this port.

---

> Native vlan is still 1 (this a configuration of the bridge itself, not the ports of the bridge).

In this bridge cni, I think we never modify the bridge's vlan config.
but only port's vlan config.
So the native vlan that I mentioned before is always port's native vlan.

---

> With the above config, untagged ingress traffic will have tag 100 (clearly not the native vlan) and on egress, vid 100 and untagged traffic will output untagged.

I'm not really confident that I understand this part.
ingress is ok, follow PVID config.
egress part.
If we have two port eth0 & eth1 and those ports are bridge ports.
```
eth0: 1 pvid egress untagged
eth1: 100 pvid egress untagged
      1-99
```
traffic from eth0 to eth1 will be captured with vid 1 tagged in eth1 egress direction.

The only config with problem will be like this

```
eth0: 1 pvid egress untagged
eth1: 1 egress untagged
      100 pvid egress untagged
```
traffic from eth0 to eth1 will be captured with untagged in eth1 egress direction.
but traffic from eth1 to eth0 will be catpured with tagged vid 100 in eth0 egress direction (so drop.)

---

> I'm familiar with cisco switches as well and I don't remember a config to do this.
This is my config of my switch.
Cisco didn't have "bridge's vlan config" in Linux.
cisco only have port's vlan config.
So I'm not sure which config that is not in cisco switch.

c2960s
```
interface GigabitEthernet1/0/48
 description to-4100
 switchport trunk native vlan 101
 switchport trunk allowed vlan 100,101,999
 switchport mode trunk
!
```

---

So from my side, I will suggest to remove default vlan always when vlanID is presented.

Stay away from the config as this 
```
// Bad 
1 Egress untagged
100 PVID Egress untagged

// Good (vlanID == 100)
100 PVID Egress untagged

// Good (vlanID == 0)
1 PVID Egress untagged
```

**Comment by tjjh89017** [comment]
If we took this cisco config as example to impelment.
```
interface GigabitEthernet1/0/48
 description to-4100
 switchport trunk native vlan 101
 switchport trunk allowed vlan 100,101,999
 switchport mode trunk
!
```

```
vlanID := 101
vlanTrunk := [{ID: 100}, {ID: 101}, {ID: 999}]
```

In Linux Bridge, we should get something like this below.

```
veth0 101 PVID Egress Untagged
      100
      999
```
without `1 Egress Untagged` inside.

I want to ensure that we have the same config translation in this Cisco config.

**Comment by mlguerrero12** [comment]
If you want to achieve

veth0 101 PVID Egress Untagged
      100
      999

you need to create a bridge with a default vlan equal to 101 and then add tags 100 and 999 with this pr.

In my view, you want to achieve the same with a hack. You want to add a vlan with PVID (access port) 101 and then delete the default vlan (with preserveDefaultVlan equal to false) and add tags 100 and 999.

Conceptually, this is not correct.

**Comment by mlguerrero12** [comment]
All the issues come from not having removed the default vlan of the bridge from the ports since day 1.

If we allow users this kind of flexibility, there will be confusion. They can set a different default vlan by setting a PVID and having preserveDefaultVlan equal to false. But this is contradicting, they will have a default vlan but they are explicitly saying that they don't want to preserve the default vlan. It works, but it's not semantically correct.

Another issue, if they leave the default vlan of the bridge and set a PVID. They will have on port:
1 Egress Untagged
101 PVID Untagged

Untagged ingress traffic is fine, it will be tagged with 101, but egress traffic with tag 101 and 1 will be untagged. In a cisco switch, if the default vlan is set to 101, the trunk will not modify any other traffic. Traffic with vlan 1 will have tag 1. This will not be the case here.

In my opinion, we should stick to basic concepts, at least for now to avoid these issues. A trunk port should only allow tagged vids and the native vlan, the one configured on the bridge.

@maiqueb @squeed, what do you think? See my comment above also.

**Comment by tjjh89017** [comment]
> All the issues come from not having removed the default vlan of the bridge from the ports since day 1.
> 
> If we allow users this kind of flexibility, there will be confusion. They can set a different default vlan by setting a PVID and having preserveDefaultVlan equal to false. But this is contradicting, they will have a default vlan but they are explicitly saying that they don't want to preserve the default vlan. It works, but it's not semantically correct.
> 
> Another issue, if they leave the default vlan of the bridge and set a PVID. They will have on port: 1 Egress Untagged 101 PVID Untagged
> 
> Untagged ingress traffic is fine, it will be tagged with 101, but egress traffic with tag 101 and 1 will be untagged. In a cisco switch, if the default vlan is set to 101, the trunk will not modify any other traffic. Traffic with vlan 1 will have tag 1. This will not be the case here.
> 
> In my opinion, we should stick to basic concepts, at least for now to avoid these issues. A trunk port should only allow tagged vids and the native vlan, the one configured on the bridge.
> 
> @maiqueb @squeed, what do you think? See my comment above also.

As discussion with @mlguerrero12, I think we could make this PR with some limitation and leave something to the future version.
and I will file a ticket to deprecate some current behavior with breaking change to make sure there is no conflict in network scope and vlan/pvid/vids usage.

For now, I will modify the PR to support access port (untagged only) and trunk port (tagged only).
Leaving hybrid port support in the future.

---

by the way, I will suggest not to use the "default" vlan term.
Even Linux bridge could do some weird config like this [1]
But in current CNI config design, we only allow/accept common switch style config.
I will leave this to the next ticket to fix up and probably we need to remove `preserveDefaultVlan` option in the future.


[1]:
```
1 Egress Untagged
101 PVID Untagged
```

**Comment by mlguerrero12** [comment]
cool, thanks @tjjh89017

File: plugins/main/bridge/bridge.go (hunk 3 above)
Comment (mlguerrero12) [reply plugins/main/bridge/bridge.go:192]: ignore my last comment, you latest fix is ok

File: plugins/main/bridge/bridge.go [hunk 7]
```diff
@@ -127,6 +142,58 @@ func loadNetConf(bytes []byte, envArgs string) (*NetConf, string, error) {
...
+			if maxID <= 0 || maxID > 4094 {
+				return nil, errors.New("incorrect trunk maxID parameter")
+			}
+			if maxID < minID {
+				return nil, errors.New("minID is greater than maxID in trunk parameter")
+			}
+			for v := minID; v <= maxID; v++ {
+				vlanMap[v] = struct{}{}
+			}
+		} else if item.MinID != nil || item.MaxID != nil {
```
Comment (maiqueb) [inline plugins/main/bridge/bridge.go:172]: isn't this clause wrong ? Shouldn't we be looking if one of them is nil instead ? 

And we probably can print exactly what is missing in the error, instead of saying one of them is missing.

File: plugins/main/bridge/bridge.go (hunk 2 above)
Comment (maiqueb) [reply plugins/main/bridge/bridge.go:178]: all good.

**Review by maiqueb — CHANGES_REQUESTED** [review]
Would it make sense to also unit-test the `collectVlanTrunk` function ? Especially it's "negative" scenario ? (e.g. check a trunk without maxID but with minID, etc).

File: plugins/main/bridge/bridge.go (hunk 7 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:172]: hmm ok, let me split it to 3 conditions.

**Comment by maiqueb** [comment]
> All the issues come from not having removed the default vlan of the bridge from the ports since day 1.
> 
> If we allow users this kind of flexibility, there will be confusion. They can set a different default vlan by setting a PVID and having preserveDefaultVlan equal to false. But this is contradicting, they will have a default vlan but they are explicitly saying that they don't want to preserve the default vlan. It works, but it's not semantically correct.
> 
> Another issue, if they leave the default vlan of the bridge and set a PVID. They will have on port: 1 Egress Untagged 101 PVID Untagged
> 
> Untagged ingress traffic is fine, it will be tagged with 101, but egress traffic with tag 101 and 1 will be untagged. In a cisco switch, if the default vlan is set to 101, the trunk will not modify any other traffic. Traffic with vlan 1 will have tag 1. This will not be the case here.
> 
> In my opinion, we should stick to basic concepts, at least for now to avoid these issues. A trunk port should only allow tagged vids and the native vlan, the one configured on the bridge.
> 
> @maiqueb @squeed, what do you think? See my comment above also.

I agree w/ @mlguerrero12 - having a single, crystal clear flow regarding VLANs is preferable.

**Comment by tjjh89017** [comment]
From my side,

> A trunk port should only allow tagged vids and the native vlan, the one configured on the bridge.

I didn't agree with this one in the future.
But currently, it's ok.

File: plugins/main/bridge/bridge.go [hunk 8]
```diff
@@ -174,9 +175,9 @@ func collectVlanTrunk(vlanTrunk []*VlanTrunk) ([]int, error) {
 			for v := minID; v <= maxID; v++ {
 				vlanMap[v] = struct{}{}
 			}
-		} else if item.MinID == nil && item.MaxID != nil {
+		case item.MinID == nil && item.MaxID != nil:
```
Comment (maiqueb) [inline plugins/main/bridge/bridge.go:178]: nit: maybe omit the != nil ? you know one of them is empty ; just checking for the empty ones would be OK, right ?

File: plugins/main/bridge/bridge.go [hunk 9]
```diff
@@ -174,9 +175,9 @@ func collectVlanTrunk(vlanTrunk []*VlanTrunk) ([]int, error) {
...
 			return nil, errors.New("minID and maxID should be configured simultaneously, minID is missing")
-		} else if item.MinID != nil && item.MaxID == nil {
+		case item.MinID != nil && item.MaxID == nil:
```
Comment (maiqueb) [inline plugins/main/bridge/bridge.go:180]: ditto

File: plugins/main/bridge/bridge.go [hunk 10]
```diff
@@ -372,18 +447,27 @@ func setupVeth(netns ns.NetNS, br *netlink.Bridge, ifName string, mtu int, hairp
...
 		}
+	}
 
+	// Currently bridge CNI only support access port(untagged only) or trunk port(tagged only)
+	if vlanID != 0 {
 		err = netlink.BridgeVlanAdd(hostVeth, uint16(vlanID), true, true, false, true)
 		if err != nil {
 			return nil, nil, fmt.Errorf("failed to setup vlan tag on interface %q: %v", hostIface.Name, err)
 		}
+	} else {
```
Comment (mlguerrero12) [inline plugins/main/bridge/bridge.go:463]: you don't need this else

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:463]: because we block the condition in the early condition?

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go]: Ok remove it.

**Review by maiqueb — COMMENTED** [review]
just some opinionated nits.

File: plugins/main/bridge/bridge.go (hunk 8 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:178]: I will consider to add detail condition to make sure others could read it and understand it immediately.

Comment (mlguerrero12) [reply plugins/main/bridge/bridge.go:178]: the 2 of them could be empty, the != nil is needed I think

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:178]: yep, because it's switch-case
i'm not so sure about that.

Comment (maiqueb) [reply plugins/main/bridge/bridge.go:178]: we would need more unit tests then, to assert this works as intended :)

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:178]: I take a look in test.go
but I'm not sure how to add a unit test without in cmdAddDel Test case hmm.

Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:178]: all test is via cmdAddDelTest.
So it's not easy to insert an unit test for collectVlanTrunk.

Comment (maiqueb) [reply plugins/main/bridge/bridge.go:178]: I would just add a new context in the describe block and have something like:

```golang

DescribeTable(
    "collectVlanTrunk succeeds", 
    func(vlanTrunks ...*vlanTrunk, expectedVIDs []int) {
        Expect(vlancollectVlanTrunk(vlanTrunks)).To(Equal(expectedVIDs))
    }, 
    Entry("", &vlanTrunk{}, []int{}),
    Entry("", &vlanTrunk{minID: 100, maxID: 105}, []int{100, 101, 102, 103, 104, 105}),
...
)

DescribeTable(
    "collectVlanTrunk **fails**", 
    func(vlanTrunks ...*vlanTrunk, expectedError error) {
        Expect(vlancollectVlanTrunk(vlanTrunks)).To(MatchError(expectedError))
    }, 
    Entry("", &vlanTrunk{minID: 100}, fmt.Errorf("minID and maxID should be configured simultaneously, maxID is missing")),
    Entry("", &vlanTrunk{maxID: 100}, "minID and maxID should be configured simultaneously, minID is missing"),
    Entry("", &vlanTrunk{minID: -100, maxID: 105}, "incorrect trunk minID parameter"),
    Entry("", &vlanTrunk{minID: 10000, maxID: 105}, "incorrect trunk minID parameter"),
    Entry("", &vlanTrunk{minID: 100, maxID: 10000}, "incorrect trunk minID parameter"),
    Entry("", &vlanTrunk{minID: 100, maxID: -1}, "incorrect trunk minID parameter"),
...
...
)

```

File: plugins/main/bridge/bridge_test.go [hunk 11]
```diff
@@ -1672,6 +1764,25 @@ var _ = Describe("bridge Operations", func() {
...
+		correctMaxID int = 105
+	)
+
+	DescribeTable(
+		"collectVlanTrunk succeeds",
+		func(vlanTrunks []*VlanTrunk, expectedVIDs []int) {
+			Expect(collectVlanTrunk(vlanTrunks)).To(Equal(expectedVIDs))
+		},
+		Entry("", &[]*VlanTrunk{}, []int{}),
+		Entry("", &[]*VlanTrunk{
```
Comment (maiqueb) [inline plugins/main/bridge/bridge_test.go:1778]: please fill the entry descriptions; otherwise, if the tests fail, we will not know which of them has failed.

**Review by maiqueb — CHANGES_REQUESTED** [review]
missing descriptions in the table `Entry`s and also missing the "error" scenario for checking `collectVlanTrunk` fails when provided wrong input.

File: plugins/main/bridge/bridge_test.go (hunk 11 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge_test.go:1778]: I'm trial and error right now.
I'm not sure the reason I cannot test it on my PC.
So I'm asking github action to help.
Please hold on.

Comment (tjjh89017) [reply plugins/main/bridge/bridge_test.go:1778]: Sorry to bother.
I used another branch to test right now.

Comment (tjjh89017) [reply plugins/main/bridge/bridge_test.go:1778]: I used my repo to run action.
so it works to test.

Comment (maiqueb) [inline plugins/main/bridge/bridge_test.go:1778]: The type of vlanTrunk is wrong.

```suggestion
		Entry("", []*VlanTrunk{}, []int{}),
		Entry("", []*VlanTrunk{
```

File: plugins/main/bridge/bridge_test.go [hunk 12]
```diff
@@ -1672,6 +1764,25 @@ var _ = Describe("bridge Operations", func() {
...
 
+	var (
+		correctMinID int = 100
+		correctMaxID int = 105
+	)
+
+	DescribeTable(
+		"collectVlanTrunk succeeds",
+		func(vlanTrunks []*VlanTrunk, expectedVIDs []int) {
+			Expect(collectVlanTrunk(vlanTrunks)).To(Equal(expectedVIDs))
```
Comment (maiqueb) [inline plugins/main/bridge/bridge_test.go:1775]:
```suggestion
			Expect(collectVlanTrunk(vlanTrunks)).To(ConsistOf(expectedVIDs))
```

otherwise you get the following error:
```
  [FAILED] Expected
      <[]int | len:0, cap:0>: nil
  to equal
      <[]int | len:0, cap:0>: []
  In [It] at: /home/mduarted/github/cni-plugins/plugins/main/bridge/bridge_test.go:1775
```

**Review by maiqueb — CHANGES_REQUESTED** [review]

**Comment by tjjh89017** [comment]
@maiqueb I think this version of unit test should work.
Please help me to review
Thanks a lot!

File: plugins/main/bridge/bridge_test.go (hunk 12 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge_test.go]: Thanks for advice
this one is better.

Comment (maiqueb) [reply plugins/main/bridge/bridge_test.go]: what you had (ensure the return val is `nil` is also OK). You can use it if you prefer.

File: plugins/main/bridge/bridge_test.go [hunk 13]
```diff
@@ -1672,6 +1764,41 @@ var _ = Describe("bridge Operations", func() {
...
+			_, err := collectVlanTrunk(vlanTrunks)
+			Expect(err).To(MatchError(expectedError))
+		},
+		Entry("when not passed the maxID", []*VlanTrunk{{MinID: &correctMinID}}, fmt.Errorf("minID and maxID should be configured simultaneously, maxID is missing")),
+		Entry("when not passed the minID", []*VlanTrunk{{MaxID: &correctMaxID}}, fmt.Errorf("minID and maxID should be configured simultaneously, minID is missing")),
+		Entry("when the minID is negative", []*VlanTrunk{{MinID: &negativeID, MaxID: &correctMaxID}}, fmt.Errorf("incorrect trunk minID parameter")),
+		Entry("when the minID is larger than 4094", []*VlanTrunk{{MinID: &overID, MaxID: &correctMaxID}}, fmt.Errorf("incorrect trunk minID parameter")),
+		Entry("when the maxID is larger than 4094", []*VlanTrunk{{MinID: &correctMinID, MaxID: &overID}}, fmt.Errorf("incorrect trunk maxID parameter")),
+		Entry("when the maxID is negative", []*VlanTrunk{{MinID: &correctMinID, MaxID: &overID}}, fmt.Errorf("incorrect trunk maxID parameter")),
+	)
```
Comment (maiqueb) [inline plugins/main/bridge/bridge_test.go:1800]: please add a test where the maxID < minID to assert that error.

File: plugins/main/bridge/bridge.go (hunk 8 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge.go:178]: Unit tests are included in latest commit

**Review by maiqueb — CHANGES_REQUESTED** [review]
Nice, I'm just missing a few tests.

Also missing tests for the `vlanTrunk`.ID attribute; I would say:
- positive tests: a vlan trunk w/ just the ID
- positive tests: a vlan trunk w/ a range + ID
- negative test: a a vlan trunk where the ID has a negative value
- negative test: a a vlan trunk where the ID == 10000

File: plugins/main/bridge/bridge_test.go (hunk 13 above)
Comment (tjjh89017) [reply plugins/main/bridge/bridge_test.go:1800]: got it.

Comment (tjjh89017) [reply plugins/main/bridge/bridge_test.go:1820]: added more tests.

**Review by maiqueb — APPROVED** [review]
Thank you. 

Please squash all commits into one once the PR is ready. 

@mlguerrero12 can you take another look ?

**Comment by tjjh89017** [comment]
> Thank you.
> 
> Please squash all commits into one once the PR is ready.
> 
> @mlguerrero12 can you take another look ?

I will squash commit in the last moment.
Please notify me if all reviewers accept and it's ready to merge.
Thanks a lot.

**Comment by mlguerrero12** [comment]
thank you @tjjh89017

**Comment by tjjh89017** [comment]
Ok thanks all of you guys!
I already squashed all commits into one.
Please check it.
Thanks again

**Comment by tjjh89017** [comment]
> please don't forget to update the docs, the example sets both vlan and vlantrunk

Sure, I will!

**Review by squeed — APPROVED** [review]
